File: modules/guardian-node/jaeger/threat_hint_v2_transport.py
"""Local-only canonical ThreatHint v2 transport payload framing."""

# Exact built-in types and fixed framing are protocol requirements.
# pylint: disable=too-many-instance-attributes,unidiomatic-typecheck

from __future__ import annotations

import json
import re
import weakref
from dataclasses import dataclass
from typing import Any, Dict, Iterable

from jaeger.threat_hint_v2_proof_envelope import (
    MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES,
    ThreatHintV2ProofEnvelope,
    ThreatHintV2ProofEnvelopeError,
)
from jaeger.observable_approval import (
    MAX_APPROVAL_LIFETIME_SECONDS,
    MAX_CANONICAL_APPROVAL_BYTES,
)
from jaeger.threat_observable import (
    MAX_CANONICAL_BYTES,
    ObservableBundle,
    validate_network_id,
)

TRANSPORT_MAGIC = b"PHT2"
TRANSPORT_VERSION = 1
REPORT_NONCE_BYTES = 32
LENGTH_FIELD_BYTES = 4
HEADER_BYTES = len(TRANSPORT_MAGIC) + 1 + REPORT_NONCE_BYTES + 3 * LENGTH_FIELD_BYTES
NONCE_OFFSET = len(TRANSPORT_MAGIC) + 1
ENVELOPE_LEN_OFFSET = NONCE_OFFSET + REPORT_NONCE_BYTES
BUNDLE_LEN_OFFSET = ENVELOPE_LEN_OFFSET + LENGTH_FIELD_BYTES
APPROVAL_LEN_OFFSET = BUNDLE_LEN_OFFSET + LENGTH_FIELD_BYTES

MAX_TRANSPORT_ENVELOPE_BYTES = MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES
MAX_TRANSPORT_BUNDLE_BYTES = MAX_CANONICAL_BYTES
MAX_TRANSPORT_APPROVAL_BYTES = MAX_CANONICAL_APPROVAL_BYTES
MAX_TRANSPORT_PAYLOAD_BYTES = (
    HEADER_BYTES
    + MAX_TRANSPORT_ENVELOPE_BYTES
    + MAX_TRANSPORT_BUNDLE_BYTES
    + MAX_TRANSPORT_APPROVAL_BYTES
)

APPROVAL_SHAPE_SCHEMA_VERSION = 1
APPROVAL_SHAPE_PURPOSE = "guardian_analysis_v1"
UINT64_MAX = (1 << 64) - 1


class ThreatHintV2TransportError(ValueError):
    """Redacted failure for every invalid payload or trusted network."""

    def __init__(self) -> None:
        super().__init__("invalid threat-hint v2 transport payload")


@dataclass(frozen=True, init=False, repr=False, eq=False)
class ThreatHintV2TransportPayload:
    """Canonical local v2 transport payload with no authority of its own.

    The payload carries exactly three untrusted candidate wires (a canonical
    v2 proof envelope, a canonical observable bundle, and a canonical
    approval). It never carries a relation manifest, current time, trusted
    network, authority, policy, recipient scope, or any signer key.

    The report nonce is an UNTRUSTED session lookup key only: parsing checks
    that it equals the envelope statement nonce for internal consistency, but
    equality with attacker-controlled bytes grants nothing. Downstream
    consumers must resolve the nonce against separately trusted active local
    state before calling promotion.

    The nested approval passed a canonical shape check only. Trusted key and
    recipient-scope equality, the current-time validity window, bundle
    commitment binding, and Schnorr signature verification all remain
    downstream; shape validity is not approval.
    """

    report_nonce: bytes
    envelope_wire: bytes
    bundle_wire: bytes
    approval_wire: bytes
    parsed_envelope: ThreatHintV2ProofEnvelope
    parsed_bundle: ObservableBundle

    def __init__(self) -> None:
        """Reject direct construction outside canonical parsing."""
        raise TypeError("direct transport payload construction is disabled")

    @classmethod
    def parse_canonical(
        cls, wire_bytes: bytes, trusted_network_id: str
    ) -> "ThreatHintV2TransportPayload":
        """Parse exact framing against a separately trusted local network."""
        if cls is not ThreatHintV2TransportPayload or type(wire_bytes) is not bytes:
            raise ThreatHintV2TransportError()
        if type(trusted_network_id) is not str:
            raise ThreatHintV2TransportError()
        if (
            len(wire_bytes) == 0
            or len(wire_bytes) > MAX_TRANSPORT_PAYLOAD_BYTES
            or len(wire_bytes) < HEADER_BYTES + 3
        ):
            raise ThreatHintV2TransportError()
        if (
            wire_bytes[: len(TRANSPORT_MAGIC)] != TRANSPORT_MAGIC
            or wire_bytes[NONCE_OFFSET - 1] != TRANSPORT_VERSION
        ):
            raise ThreatHintV2TransportError()

        report_nonce = wire_bytes[NONCE_OFFSET : NONCE_OFFSET + REPORT_NONCE_BYTES]
        envelope_len = _read_be_len(wire_bytes, ENVELOPE_LEN_OFFSET)
        bundle_len = _read_be_len(wire_bytes, BUNDLE_LEN_OFFSET)
        approval_len = _read_be_len(wire_bytes, APPROVAL_LEN_OFFSET)
        if (
            not 1 <= envelope_len <= MAX_TRANSPORT_ENVELOPE_BYTES
            or not 1 <= bundle_len <= MAX_TRANSPORT_BUNDLE_BYTES
            or not 1 <= approval_len <= MAX_TRANSPORT_APPROVAL_BYTES
        ):
            raise ThreatHintV2TransportError()
        if HEADER_BYTES + envelope_len + bundle_len + approval_len != len(wire_bytes):
            raise ThreatHintV2TransportError()

        bundle_start = HEADER_BYTES + envelope_len
        approval_start = bundle_start + bundle_len
        envelope_wire = wire_bytes[HEADER_BYTES:bundle_start]
        bundle_wire = wire_bytes[bundle_start:approval_start]
        approval_wire = wire_bytes[approval_start:]

        try:
            envelope = ThreatHintV2ProofEnvelope.parse_canonical(
                envelope_wire, trusted_network_id
            )
        except ThreatHintV2ProofEnvelopeError:
            raise ThreatHintV2TransportError() from None
        if envelope.parsed_statement.report_nonce != report_nonce.hex():
            raise ThreatHintV2TransportError()

        try:
            bundle = ObservableBundle.parse_canonical(bundle_wire)
        except ValueError:
            raise ThreatHintV2TransportError() from None

        _validate_approval_shape(approval_wire)

        payload = object.__new__(cls)
        object.__setattr__(payload, "report_nonce", report_nonce)
        object.__setattr__(payload, "envelope_wire", envelope_wire)
        object.__setattr__(payload, "bundle_wire", bundle_wire)
        object.__setattr__(payload, "approval_wire", approval_wire)
        object.__setattr__(payload, "parsed_envelope", envelope)
        object.__setattr__(payload, "parsed_bundle", bundle)
        _PARSED_CANONICAL[payload] = wire_bytes
        if payload.canonical_bytes != wire_bytes:
            raise ThreatHintV2TransportError()
        return payload

    @property
    def canonical_bytes(self) -> bytes:
        """Return the revalidated exact canonical framing bytes."""
        canonical_at_parse = _PARSED_CANONICAL.get(self)
        if canonical_at_parse is None:
            raise ThreatHintV2TransportError()
        self._validate_state()
        canonical = b"".join(
            [
                TRANSPORT_MAGIC,
                bytes([TRANSPORT_VERSION]),
                self.report_nonce,
                len(self.envelope_wire).to_bytes(LENGTH_FIELD_BYTES, "big"),
                len(self.bundle_wire).to_bytes(LENGTH_FIELD_BYTES, "big"),
                len(self.approval_wire).to_bytes(LENGTH_FIELD_BYTES, "big"),
                self.envelope_wire,
                self.bundle_wire,
                self.approval_wire,
            ]
        )
        if (
            len(canonical) == 0
            or len(canonical) > MAX_TRANSPORT_PAYLOAD_BYTES
            or canonical != canonical_at_parse
        ):
            raise ThreatHintV2TransportError()
        return canonical

    def _validate_state(self) -> None:
        """Revalidate private state before serialization."""
        if type(self) is not ThreatHintV2TransportPayload:
            raise ThreatHintV2TransportError()
        for wire, cap in (
            (self.envelope_wire, MAX_TRANSPORT_ENVELOPE_BYTES),
            (self.bundle_wire, MAX_TRANSPORT_BUNDLE_BYTES),
            (self.approval_wire, MAX_TRANSPORT_APPROVAL_BYTES),
        ):
            if type(wire) is not bytes or not 1 <= len(wire) <= cap:
                raise ThreatHintV2TransportError()
        if type(self.report_nonce) is not bytes:
            raise ThreatHintV2TransportError()
        if len(self.report_nonce) != REPORT_NONCE_BYTES:
            raise ThreatHintV2TransportError()
        if type(self.parsed_envelope) is not ThreatHintV2ProofEnvelope:
            raise ThreatHintV2TransportError()
        if type(self.parsed_bundle) is not ObservableBundle:
            raise ThreatHintV2TransportError()
        if self.parsed_envelope.canonical_bytes != self.envelope_wire:
            raise ThreatHintV2TransportError()
        if self.parsed_bundle.canonical_bytes != self.bundle_wire:
            raise ThreatHintV2TransportError()
        if (
            self.parsed_envelope.parsed_statement.report_nonce
            != self.report_nonce.hex()
        ):
            raise ThreatHintV2TransportError()
        _validate_approval_shape(self.approval_wire)


def _read_be_len(wire_bytes: bytes, offset: int) -> int:
    return int.from_bytes(
        wire_bytes[offset : offset + LENGTH_FIELD_BYTES], byteorder="big", signed=False
    )


def _validate_approval_shape(wire_bytes: bytes) -> None:
    """Apply the canonical approval shape check possible without context.

    Trusted key equality, recipient-scope equality, the current-time validity
    window, bundle commitment binding, and Schnorr signature verification all
    remain downstream in ``verify_observable_approval``.
    """
    if type(wire_bytes) is not bytes or not (
        0 < len(wire_bytes) <= MAX_TRANSPORT_APPROVAL_BYTES
    ):
        raise ThreatHintV2TransportError()

    try:
        decoded = json.loads(
            wire_bytes.decode("utf-8"), object_pairs_hook=_reject_duplicate_keys
        )
    except (UnicodeDecodeError, ValueError):
        raise ThreatHintV2TransportError() from None
    if type(decoded) is not dict or set(decoded.keys()) != {
        "schema_version",
        "observable_commitment",
        "approver_xonly_public_key",
        "purpose",
        "recipient_scope",
        "network_id",
        "not_before",
        "expires_at",
        "approval_nonce",
        "signature",
    }:
        raise ThreatHintV2TransportError()

    canonical = json.dumps(
        {
            "schema_version": decoded["schema_version"],
            "observable_commitment": decoded["observable_commitment"],
            "approver_xonly_public_key": decoded["approver_xonly_public_key"],
            "purpose": decoded["purpose"],
            "recipient_scope": decoded["recipient_scope"],
            "network_id": decoded["network_id"],
            "not_before": decoded["not_before"],
            "expires_at": decoded["expires_at"],
            "approval_nonce": decoded["approval_nonce"],
            "signature": decoded["signature"],
        },
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")
    if canonical != wire_bytes:
        raise ThreatHintV2TransportError()

    if (
        type(decoded["schema_version"]) is not int
        or decoded["schema_version"] != APPROVAL_SHAPE_SCHEMA_VERSION
        or type(decoded["purpose"]) is not str
        or decoded["purpose"] != APPROVAL_SHAPE_PURPOSE
    ):
        raise ThreatHintV2TransportError()
    if type(decoded["network_id"]) is not str:
        raise ThreatHintV2TransportError()
    for field in ("not_before", "expires_at"):
        if type(decoded[field]) is not int or decoded[field] > UINT64_MAX:
            raise ThreatHintV2TransportError()

    lifetime = decoded["expires_at"] - decoded["not_before"]
    if (
        decoded["not_before"] <= 0
        or lifetime <= 0
        or lifetime > MAX_APPROVAL_LIFETIME_SECONDS
    ):
        raise ThreatHintV2TransportError()

    try:
        validate_network_id(decoded["network_id"])
    except ValueError:
        raise ThreatHintV2TransportError() from None

    for field, hex_len in (
        ("observable_commitment", 64),
        ("approver_xonly_public_key", 64),
        ("recipient_scope", 64),
        ("approval_nonce", 64),
        ("signature", 128),
    ):
        value = decoded[field]
        if (
            type(value) is not str
            or len(value) != hex_len
            or _LOWER_HEX_RE.fullmatch(value) is None
        ):
            raise ThreatHintV2TransportError()


def _reject_duplicate_keys(items: Iterable[tuple[str, Any]]) -> Dict[str, Any]:
    result: Dict[str, Any] = {}
    for key, value in items:
        if key in result:
            raise ThreatHintV2TransportError()
        result[key] = value
    return result


_LOWER_HEX_RE = re.compile(r"[0-9a-f]*$")
_PARSED_CANONICAL: weakref.WeakKeyDictionary[Any, bytes] = weakref.WeakKeyDictionary()

File: modules/threat-hint/src/threat_hint_v2_transport.rs
//! Local-only canonical ThreatHint v2 transport payload framing.
//!
//! The payload carries exactly three untrusted candidate wires: a canonical v2
//! proof envelope, a canonical observable bundle, and a canonical approval. It
//! never carries a relation manifest, current time, trusted network,
//! authority, policy, recipient scope, or any signer key.
//!
//! The 32-byte report nonce is an UNTRUSTED session lookup key only: the
//! constructor checks that it equals the envelope statement nonce for internal
//! consistency, but equality with attacker-controlled bytes grants nothing.
//! Downstream consumers must resolve the nonce against separately trusted
//! active local state before calling promotion.
//!
//! The nested approval receives a canonical shape check only, because full
//! approval verification needs a separately trusted key, recipient scope,
//! current time, and network context. Final approval verification remains
//! downstream via `verify_observable_approval`; shape validity is not
//! approval.

use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::observable_bundle::validate_network_id;
use crate::{
    ObservableBundle, ThreatHintV2ProofEnvelope, MAX_APPROVAL_LIFETIME_SECONDS,
    MAX_CANONICAL_APPROVAL_BYTES, MAX_CANONICAL_OBSERVABLE_BUNDLE_BYTES,
    MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES,
};

/// Fixed transport payload magic bytes.
pub const THREAT_HINT_V2_TRANSPORT_MAGIC: [u8; 4] = *b"PHT2";
/// The only supported transport payload framing version.
pub const THREAT_HINT_V2_TRANSPORT_VERSION: u8 = 1;
/// Exact untrusted report nonce length in bytes.
pub const REPORT_NONCE_BYTES: usize = 32;
/// Maximum nested canonical v2 proof envelope wire size.
pub const MAX_TRANSPORT_ENVELOPE_BYTES: usize = MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES;
/// Maximum nested canonical observable bundle wire size.
pub const MAX_TRANSPORT_BUNDLE_BYTES: usize = MAX_CANONICAL_OBSERVABLE_BUNDLE_BYTES;
/// Maximum nested canonical approval wire size.
pub const MAX_TRANSPORT_APPROVAL_BYTES: usize = MAX_CANONICAL_APPROVAL_BYTES;

const LENGTH_FIELD_BYTES: usize = 4;
const HEADER_BYTES: usize =
    THREAT_HINT_V2_TRANSPORT_MAGIC.len() + 1 + REPORT_NONCE_BYTES + 3 * LENGTH_FIELD_BYTES;
const NONCE_OFFSET: usize = THREAT_HINT_V2_TRANSPORT_MAGIC.len() + 1;
const ENVELOPE_LEN_OFFSET: usize = NONCE_OFFSET + REPORT_NONCE_BYTES;
const BUNDLE_LEN_OFFSET: usize = ENVELOPE_LEN_OFFSET + LENGTH_FIELD_BYTES;
const APPROVAL_LEN_OFFSET: usize = BUNDLE_LEN_OFFSET + LENGTH_FIELD_BYTES;

/// Maximum accepted transport payload size; the total length is exact.
pub const MAX_TRANSPORT_PAYLOAD_BYTES: usize = HEADER_BYTES
    + MAX_TRANSPORT_ENVELOPE_BYTES
    + MAX_TRANSPORT_BUNDLE_BYTES
    + MAX_TRANSPORT_APPROVAL_BYTES;

const APPROVAL_SHAPE_SCHEMA_VERSION: u16 = 1;
const APPROVAL_SHAPE_PURPOSE: &str = "guardian_analysis_v1";
const FIXED_HASH_HEX_LEN: usize = 64;
const SIGNATURE_HEX_LEN: usize = 128;

/// Redacted failure returned for every invalid payload or trusted network.
#[derive(Debug, Error, PartialEq, Eq)]
pub enum ThreatHintV2TransportError {
    /// The payload, a nested wire, or the separately trusted network is invalid.
    #[error("invalid threat-hint v2 transport payload")]
    InvalidPayload,
}

/// A canonical local ThreatHint v2 transport payload.
///
/// Direct construction is unavailable; callers must use
/// [`ThreatHintV2TransportPayload::parse_canonical`] with a separately trusted
/// network. The report nonce and all nested wires remain untrusted input:
/// structural validity grants no proof, approval, disclosure, or promotion
/// authority.
#[derive(Clone, PartialEq, Eq)]
pub struct ThreatHintV2TransportPayload {
    report_nonce: [u8; REPORT_NONCE_BYTES],
    envelope_wire: Vec<u8>,
    bundle_wire: Vec<u8>,
    approval_wire: Vec<u8>,
    envelope: ThreatHintV2ProofEnvelope,
    bundle: ObservableBundle,
}

impl core::fmt::Debug for ThreatHintV2TransportPayload {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("ThreatHintV2TransportPayload")
            .finish_non_exhaustive()
    }
}

#[derive(Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
struct ApprovalShapeWire {
    schema_version: u16,
    observable_commitment: String,
    approver_xonly_public_key: String,
    purpose: String,
    recipient_scope: String,
    network_id: String,
    not_before: u64,
    expires_at: u64,
    approval_nonce: String,
    signature: String,
}

impl ThreatHintV2TransportPayload {
    /// Parses exact canonical framing against a separately trusted network.
    ///
    /// The framing has a fixed magic and version, big-endian u32 lengths,
    /// strictly nonzero capped wires, and an exact total length with no
    /// trailing bytes. The constructor independently parses the nested
    /// envelope against `trusted_network_id`, requires the envelope statement
    /// report nonce to equal the payload nonce, parses the nested bundle as
    /// canonical, and applies a canonical shape check to the nested approval.
    pub fn parse_canonical(
        wire_bytes: &[u8],
        trusted_network_id: &str,
    ) -> Result<Self, ThreatHintV2TransportError> {
        if wire_bytes.is_empty() || wire_bytes.len() > MAX_TRANSPORT_PAYLOAD_BYTES {
            return Err(ThreatHintV2TransportError::InvalidPayload);
        }
        if wire_bytes.len() < HEADER_BYTES + 3 {
            return Err(ThreatHintV2TransportError::InvalidPayload);
        }
        if wire_bytes[..THREAT_HINT_V2_TRANSPORT_MAGIC.len()] != THREAT_HINT_V2_TRANSPORT_MAGIC
            || wire_bytes[NONCE_OFFSET - 1] != THREAT_HINT_V2_TRANSPORT_VERSION
        {
            return Err(ThreatHintV2TransportError::InvalidPayload);
        }

        let report_nonce: [u8; REPORT_NONCE_BYTES] = wire_bytes
            [NONCE_OFFSET..NONCE_OFFSET + REPORT_NONCE_BYTES]
            .try_into()
            .map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;

        let envelope_len = read_be_len(wire_bytes, ENVELOPE_LEN_OFFSET)?;
        let bundle_len = read_be_len(wire_bytes, BUNDLE_LEN_OFFSET)?;
        let approval_len = read_be_len(wire_bytes, APPROVAL_LEN_OFFSET)?;
        if envelope_len == 0
            || envelope_len > MAX_TRANSPORT_ENVELOPE_BYTES
            || bundle_len == 0
            || bundle_len > MAX_TRANSPORT_BUNDLE_BYTES
            || approval_len == 0
            || approval_len > MAX_TRANSPORT_APPROVAL_BYTES
        {
            return Err(ThreatHintV2TransportError::InvalidPayload);
        }
        if HEADER_BYTES + envelope_len + bundle_len + approval_len != wire_bytes.len() {
            return Err(ThreatHintV2TransportError::InvalidPayload);
        }

        let envelope_start = HEADER_BYTES;
        let bundle_start = envelope_start + envelope_len;
        let approval_start = bundle_start + bundle_len;
        let envelope_wire = wire_bytes[envelope_start..bundle_start].to_vec();
        let bundle_wire = wire_bytes[bundle_start..approval_start].to_vec();
        let approval_wire = wire_bytes[approval_start..].to_vec();

        let envelope =
            ThreatHintV2ProofEnvelope::parse_canonical(&envelope_wire, trusted_network_id)
                .map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;
        if envelope.statement().report_nonce_hex() != hex::encode(report_nonce) {
            return Err(ThreatHintV2TransportError::InvalidPayload);
        }

        let bundle = ObservableBundle::parse_canonical(&bundle_wire)
            .map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;

        validate_approval_shape(&approval_wire)?;

        Ok(Self {
            report_nonce,
            envelope_wire,
            bundle_wire,
            approval_wire,
            envelope,
            bundle,
        })
    }

    /// Returns the UNTRUSTED report nonce session lookup key.
    ///
    /// Callers must resolve this value against separately trusted active local
    /// state before any promotion; it is attacker-controlled input.
    pub fn report_nonce(&self) -> &[u8; REPORT_NONCE_BYTES] {
        &self.report_nonce
    }

    /// Returns the parsed nested v2 proof envelope.
    pub fn envelope(&self) -> &ThreatHintV2ProofEnvelope {
        &self.envelope
    }

    /// Returns the parsed nested observable bundle.
    pub fn bundle(&self) -> &ObservableBundle {
        &self.bundle
    }

    /// Returns the exact nested envelope wire bytes.
    pub fn envelope_wire(&self) -> &[u8] {
        &self.envelope_wire
    }

    /// Returns the exact nested bundle wire bytes.
    pub fn bundle_wire(&self) -> &[u8] {
        &self.bundle_wire
    }

    /// Returns the exact nested approval wire bytes.
    ///
    /// These bytes passed a canonical shape check only; full approval
    /// verification with a separately trusted context remains downstream.
    pub fn approval_wire(&self) -> &[u8] {
        &self.approval_wire
    }

    /// Serializes the validated payload back to exact canonical framing bytes.
    pub fn to_canonical_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(
            HEADER_BYTES
                + self.envelope_wire.len()
                + self.bundle_wire.len()
                + self.approval_wire.len(),
        );
        bytes.extend_from_slice(&THREAT_HINT_V2_TRANSPORT_MAGIC);
        bytes.push(THREAT_HINT_V2_TRANSPORT_VERSION);
        bytes.extend_from_slice(&self.report_nonce);
        for wire in [&self.envelope_wire, &self.bundle_wire, &self.approval_wire] {
            bytes.extend_from_slice(&(wire.len() as u32).to_be_bytes());
        }
        for wire in [&self.envelope_wire, &self.bundle_wire, &self.approval_wire] {
            bytes.extend_from_slice(wire);
        }
        bytes
    }
}

fn read_be_len(wire_bytes: &[u8], offset: usize) -> Result<usize, ThreatHintV2TransportError> {
    let field: [u8; LENGTH_FIELD_BYTES] = wire_bytes[offset..offset + LENGTH_FIELD_BYTES]
        .try_into()
        .map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;
    Ok(u32::from_be_bytes(field) as usize)
}

/// Applies the canonical approval shape check that is possible without a
/// trusted key, recipient scope, current time, or network context.
///
/// This rejects malformed, noncanonical, mistyped, and out-of-shape approval
/// wires. Trusted key equality, recipient-scope equality, the current-time
/// validity window, bundle commitment binding, and Schnorr signature
/// verification all remain downstream in `verify_observable_approval`.
fn validate_approval_shape(wire_bytes: &[u8]) -> Result<(), ThreatHintV2TransportError> {
    if wire_bytes.is_empty() || wire_bytes.len() > MAX_TRANSPORT_APPROVAL_BYTES {
        return Err(ThreatHintV2TransportError::InvalidPayload);
    }

    let shape: ApprovalShapeWire = serde_json::from_slice(wire_bytes)
        .map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;
    let canonical =
        serde_json::to_vec(&shape).map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;
    if canonical != wire_bytes {
        return Err(ThreatHintV2TransportError::InvalidPayload);
    }

    if shape.schema_version != APPROVAL_SHAPE_SCHEMA_VERSION
        || shape.purpose != APPROVAL_SHAPE_PURPOSE
    {
        return Err(ThreatHintV2TransportError::InvalidPayload);
    }
    validate_network_id(&shape.network_id)
        .map_err(|_| ThreatHintV2TransportError::InvalidPayload)?;

    let lifetime = shape
        .expires_at
        .checked_sub(shape.not_before)
        .ok_or(ThreatHintV2TransportError::InvalidPayload)?;
    if shape.not_before == 0 || lifetime == 0 || lifetime > MAX_APPROVAL_LIFETIME_SECONDS {
        return Err(ThreatHintV2TransportError::InvalidPayload);
    }

    if !is_lower_hex(&shape.observable_commitment, FIXED_HASH_HEX_LEN)
        || !is_lower_hex(&shape.approver_xonly_public_key, FIXED_HASH_HEX_LEN)
        || !is_lower_hex(&shape.recipient_scope, FIXED_HASH_HEX_LEN)
        || !is_lower_hex(&shape.approval_nonce, FIXED_HASH_HEX_LEN)
        || !is_lower_hex(&shape.signature, SIGNATURE_HEX_LEN)
    {
        return Err(ThreatHintV2TransportError::InvalidPayload);
    }

    Ok(())
}

fn is_lower_hex(value: &str, expected_len: usize) -> bool {
    value.len() == expected_len
        && value
            .as_bytes()
            .iter()
            .all(|byte| matches!(byte, b'0'..=b'9' | b'a'..=b'f'))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ThreatHintV2Statement;

    const NONCE_HEX: &str = "a4805ae9672576df4c3f2f60b9d99ab138206a09dacd82211cfe012e2b17dc22";

    fn statement_wire(network_id: &str) -> Vec<u8> {
        format!(
            "{{\"schema_version\":2,\"artifact_hash\":\"{}\",\"observable_commitment\":\"{}\",\"confidence_bps\":7500,\"disclosure_class\":\"review_required_v1\",\"report_nonce\":\"{NONCE_HEX}\",\"observed_at\":1800000100,\"network_id\":\"{network_id}\"}}",
            "00".repeat(32),
            "11".repeat(32),
        )
        .into_bytes()
    }

    fn envelope_wire(network_id: &str) -> Vec<u8> {
        let statement =
            ThreatHintV2Statement::parse_canonical(&statement_wire(network_id), network_id)
                .expect("valid statement");
        let digest = hex::encode(statement.statement_digest().expect("digest"));
        let statement_text = String::from_utf8(statement_wire(network_id)).expect("ASCII");
        format!(
            "{{\"schema_version\":2,\"protocol_id\":\"/prometheus/threat-hint/2.0.0\",\"relation_id\":\"prometheus-threat-hint-v2\",\"statement\":{},\"statement_digest\":\"{digest}\",\"proof\":\"{}\"}}",
            serde_json::to_string(&statement_text).expect("string"),
            "aa".repeat(16)
        )
        .into_bytes()
    }

    fn bundle_wire() -> Vec<u8> {
        br#"{"schema_version":1,"disclosure_policy":"review_required_v1","scope":{"platform":"linux","format":"elf"},"observables":[{"kind":"api_import","value":"mmap"}]}"#.to_vec()
    }

    fn approval_wire() -> Vec<u8> {
        format!(
            "{{\"schema_version\":1,\"observable_commitment\":\"{}\",\"approver_xonly_public_key\":\"{}\",\"purpose\":\"guardian_analysis_v1\",\"recipient_scope\":\"{}\",\"network_id\":\"testnet-10\",\"not_before\":1800000000,\"expires_at\":1800000600,\"approval_nonce\":\"{}\",\"signature\":\"{}\"}}",
            "5c".repeat(32),
            "86".repeat(32),
            "01".repeat(32),
            "5e".repeat(32),
            "99".repeat(64),
        )
        .into_bytes()
    }

    fn payload_wire() -> Vec<u8> {
        let nonce = hex::decode(NONCE_HEX).expect("hex");
        let envelope = envelope_wire("testnet-10");
        let bundle = bundle_wire();
        let approval = approval_wire();
        let mut bytes = Vec::new();
        bytes.extend_from_slice(&THREAT_HINT_V2_TRANSPORT_MAGIC);
        bytes.push(THREAT_HINT_V2_TRANSPORT_VERSION);
        bytes.extend_from_slice(&nonce);
        for wire in [&envelope, &bundle, &approval] {
            bytes.extend_from_slice(&(wire.len() as u32).to_be_bytes());
        }
        for wire in [&envelope, &bundle, &approval] {
            bytes.extend_from_slice(wire);
        }
        bytes
    }

    #[test]
    fn roundtrip_byte_identity_and_nonce_binding() {
        let wire = payload_wire();
        let payload =
            ThreatHintV2TransportPayload::parse_canonical(&wire, "testnet-10").expect("valid");
        assert_eq!(payload.to_canonical_bytes(), wire);
        assert_eq!(hex::encode(payload.report_nonce()), NONCE_HEX);
        assert_eq!(
            payload.envelope_wire(),
            payload
                .envelope()
                .to_canonical_bytes()
                .expect("canonical")
                .as_slice()
        );
        assert_eq!(
            payload.bundle_wire(),
            payload
                .bundle()
                .to_canonical_bytes()
                .expect("canonical")
                .as_slice()
        );
        assert_eq!(payload.envelope().statement().report_nonce_hex(), NONCE_HEX);
    }

    #[test]
    fn rejects_framing_violations() {
        let wire = payload_wire();

        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(b"", "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&wire[..30], "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let mut bad_magic = wire.clone();
        bad_magic[0] = b'X';
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&bad_magic, "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let mut bad_version = wire.clone();
        bad_version[4] = 2;
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&bad_version, "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let mut trailing = wire.clone();
        trailing.push(0x00);
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&trailing, "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let mut nonce_mismatch = wire;
        nonce_mismatch[5] ^= 0x01;
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&nonce_mismatch, "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );
    }

    #[test]
    fn rejects_wrong_trusted_network_and_bad_nested_wires() {
        let wire = payload_wire();
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&wire, "mainnet"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let envelope_len = envelope_wire("testnet-10").len();
        let bundle_start = HEADER_BYTES + envelope_len;

        let mut bad_bundle = wire.clone();
        bad_bundle[bundle_start] = b' ';
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&bad_bundle, "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let mut bad_approval = wire;
        let approval_start = bad_approval.len() - approval_wire().len();
        bad_approval[approval_start] = b' ';
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(&bad_approval, "testnet-10"),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );
    }

    #[test]
    fn approval_shape_check_rejects_context_free_violations() {
        let base = String::from_utf8(approval_wire()).expect("ASCII");

        for mutated in [
            base.replace("\"schema_version\":1", "\"schema_version\":2"),
            base.replace("guardian_analysis_v1", "guardian_analysis_v2"),
            base.replace("\"not_before\":1800000000", "\"not_before\":0"),
            base.replace("\"expires_at\":1800000600", "\"expires_at\":1800003601"),
            base.replace(&"99".repeat(64), &"99".repeat(63)),
            base.replace(&"5c".repeat(32), &"5C".repeat(32)),
            base.replace(":", ": "),
        ] {
            assert_eq!(
                validate_approval_shape(mutated.as_bytes()),
                Err(ThreatHintV2TransportError::InvalidPayload)
            );
        }

        assert!(validate_approval_shape(b"").is_err());
        assert!(validate_approval_shape(&vec![b'{'; MAX_TRANSPORT_APPROVAL_BYTES + 1]).is_err());
        assert!(validate_approval_shape(&approval_wire()).is_ok());
    }

    #[test]
    fn total_size_limit_is_fail_closed() {
        assert_eq!(
            ThreatHintV2TransportPayload::parse_canonical(
                &vec![b'P'; MAX_TRANSPORT_PAYLOAD_BYTES + 1],
                "testnet-10"
            ),
            Err(ThreatHintV2TransportError::InvalidPayload)
        );

        let wire = payload_wire();
        assert!(wire.len() < MAX_TRANSPORT_PAYLOAD_BYTES);
        assert_eq!(
            MAX_TRANSPORT_PAYLOAD_BYTES,
            HEADER_BYTES
                + MAX_TRANSPORT_ENVELOPE_BYTES
                + MAX_TRANSPORT_BUNDLE_BYTES
                + MAX_TRANSPORT_APPROVAL_BYTES
        );
    }
}
